# Clean Cars and Clean Data

The aim of this project is to take a dataset of used cars from *eBay Kleinanzeigen*, a classified section of the German eBay website, clean the data and analyze it. 

In [1]:
import numpy as np 
import pandas as pd
import re

In [2]:
autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [3]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

## A Quick Look at the Data

The imported dataset, `autos`, has 50,000 entries with 20 feature columns, 5 of which are integers and the other 15 appear to be strings or possibly datetime objects.

A few of the columns appear to be missing a few entries, namely columns:
* `vehicleType`
* `gearbox`
* `model`
* `fuelType`
* `notRepairedDamage`

In [5]:
columns = autos.columns

new_columns = []

# taken from stackoverflow (find link below)
def convert(name):
    s1 = re.sub('(.)([A-Z][a-z]+)', r'\1_\2', name)
    return re.sub('([a-z0-9])([A-Z])', r'\1_\2', s1).lower()

for col in columns:
    new_columns.append(convert(col))
    
# replace old column names with 'snakecase' column names
autos.columns = new_columns

#### NOTE:
The above function was taken from stackoverflow: [link](https://stackoverflow.com/questions/1175208/elegant-python-function-to-convert-camelcase-to-snake-case)

Remember to import regex module: `import re`

In [6]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  year_of_registration  \
0  privat    Angebot  $5,000  control          bus                  2004   
1  privat    Angebot  $8,500  control    limousine                  1997   
2  privat    Angebot  $8,990     test    limousine                  2009   
3  privat    Angebot  $4,350  control   kleinwagen                  2007   
4  privat    Angebot  $1,350     test        kombi                  2003   

     gearbox  power_ps   model   odometer  month_of_registration fuel_type  \
0    manuell       158  andere  150,000km                      3       lpg   
1  automatik       286     7er  150,000km                      6    benzin   
2    manuell       102    golf   70,000km                      7    benzin   
3  automatik        71  fortwo   70,000km                      6    benzin   
4    manuell         0   focus  150,000km                      7    benzin   

        brand not_repaired_damage         date_created  nr_of_pictures  \
0     peugeot                nein  2016-03-26 00:00:00               0   
1         bmw                nein  2016-04-04 00:00:00               0   
2  volkswagen                nein  2016-03-26 00:00:00               0   
3       smart                nein  2016-03-12 00:00:00               0   
4        ford                nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [7]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-16 21:50:53  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  year_of_registration  gearbox      power_ps  model  \
count         44905          50000.000000    47320  50000.000000  47242   
unique            8                   NaN        2           NaN    245   
top       limousine                   NaN  manuell           NaN   golf   
freq          12859                   NaN    36993           NaN   4024   
mean            NaN           2005.073280      NaN    116.355920    NaN   
std             NaN            105.712813      NaN    209.216627    NaN   
min             NaN           1000.000000      NaN      0.000000    NaN   
25%             NaN           1999.000000      NaN     70.000000    NaN   
50%             NaN           2003.000000      NaN    105.000000    NaN   
75%             NaN           2008.000000      NaN    150.000000    NaN   
max             NaN           9999.000000      NaN  17700.000000    NaN   

         odometer  month_of_registration fuel_type       brand  \
count       50000           50000.000000     45518       50000   
unique         13                    NaN         7          40   
top     150,000km                    NaN    benzin  volkswagen   
freq        32424                    NaN     30107       10687   
mean          NaN               5.723360       NaN         NaN   
std           NaN               3.711984       NaN         NaN   
min           NaN               0.000000       NaN         NaN   
25%           NaN               3.000000       NaN         NaN   
50%           NaN               6.000000       NaN         NaN   
75%           NaN               9.000000       NaN         NaN   
max           NaN              12.000000       NaN         NaN   

       not_repaired_damage         date_created  nr_of_pictures   postal_code  \
count                40171                50000         50000.0  50000.000000   
unique                   2                   76             NaN           NaN   
top                   nein  2016-04-03 00:00:00             NaN           NaN   
freq                 35232                 1946             NaN           NaN   
mean                   NaN                  NaN             0.0  50813.627300   
std                    NaN                  NaN             0.0  25779.747957   
min                    NaN                  NaN             0.0   1067.000000   
25%                    NaN                  NaN             0.0  30451.000000   
50%                    NaN                  NaN             0.0  49577.000000   
75%                    NaN                  NaN             0.0  71540.000000   
max                    NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

In [8]:
autos[autos['seller'] != 'privat']

date_crawled                                         name  \
7738  2016-03-15 18:06:22  Verkaufe_mehrere_Fahrzeuge_zum_Verschrotten   

          seller offer_type price   abtest vehicle_type  year_of_registration  \
7738  gewerblich    Angebot  $100  control        kombi                  2000   

      gearbox  power_ps   model   odometer  month_of_registration fuel_type  \
7738  manuell         0  megane  150,000km                      8    benzin   

        brand not_repaired_damage         date_created  nr_of_pictures  \
7738  renault                 NaN  2016-03-15 00:00:00               0   

      postal_code            last_seen  
7738        65232  2016-04-06 17:15:37

In [9]:
autos[autos['offer_type'] != 'Angebot']

date_crawled                  name  seller offer_type price  \
17541  2016-04-03 15:48:33  Suche_VW_T5_Multivan  privat     Gesuch    $0   

      abtest vehicle_type  year_of_registration gearbox  power_ps  \
17541   test          bus                  2005     NaN         0   

             model   odometer  month_of_registration fuel_type       brand  \
17541  transporter  150,000km                      0       NaN  volkswagen   

      not_repaired_damage         date_created  nr_of_pictures  postal_code  \
17541                 NaN  2016-04-03 00:00:00               0        29690   

                 last_seen  
17541  2016-04-05 15:16:06

In [10]:
autos[autos['nr_of_pictures'] != 0]

Empty DataFrame
Columns: [date_crawled, name, seller, offer_type, price, abtest, vehicle_type, year_of_registration, gearbox, power_ps, model, odometer, month_of_registration, fuel_type, brand, not_repaired_damage, date_created, nr_of_pictures, postal_code, last_seen]
Index: []

In [11]:
autos['abtest'].value_counts()

test       25756
control    24244
Name: abtest, dtype: int64

In [12]:
print(autos['date_crawled'].head())
print('\n')
print(type(autos['date_crawled'].head()[0]))

0    2016-03-26 17:47:46
1    2016-04-04 13:38:56
2    2016-03-26 18:57:24
3    2016-03-12 16:58:10
4    2016-04-01 14:38:50
Name: date_crawled, dtype: object


<class 'str'>


In [13]:
print(autos['price'].head())
print('\n')
print(type(autos['price'].head()[0]))

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object


<class 'str'>


In [14]:
print(autos['odometer'].head())
print('\n')
print(type(autos['odometer'].head()[0]))

0    150,000km
1    150,000km
2     70,000km
3     70,000km
4    150,000km
Name: odometer, dtype: object


<class 'str'>


In [15]:
print(autos['date_created'].head())
print('\n')
print(type(autos['date_created'].head()[0]))

0    2016-03-26 00:00:00
1    2016-04-04 00:00:00
2    2016-03-26 00:00:00
3    2016-03-12 00:00:00
4    2016-04-01 00:00:00
Name: date_created, dtype: object


<class 'str'>


In [16]:
print(autos['last_seen'].head())
print('\n')
print(type(autos['last_seen'].head()[0]))

0    2016-04-06 06:45:54
1    2016-04-06 14:45:08
2    2016-04-06 20:15:37
3    2016-03-15 03:16:28
4    2016-04-01 14:38:50
Name: last_seen, dtype: object


<class 'str'>


## A More Close Look at the Data

1. The two features `seller` and `offer_type` have the same value for all rows, `privat` (private) and `Angebot` (offer) respectively, except for one row each. Both of these rows will provide little insight into the data and can be dropped. 

2. As for the `nr_of_pictures` features, there appears to be 0 pictures for every single row and therefore can also be dropped.

3. Lastly there are two features, `price` and `odometer` which are currently encoded as strings but would be useful if converted to integers (or possibly floats).

4. Three of the features, `date_crawled`, `date_created` and `last_seen` are all in string format. It might be useful to convert these columns into datatime objects. I will holdoff on this until more analysis is performed.



In [17]:
# Drop (remove) `seller`, `offer_type` and `nr_of_pictures` 
# columns from dataframe
autos.drop(columns='seller', inplace=True)
autos.drop(columns='offer_type', inplace=True)
autos.drop(columns='nr_of_pictures', inplace=True)

In [18]:
# Clean and convert `price` column to float type
autos['price'] = autos['price'].str.replace('$', '')
autos['price'] = autos['price'].str.replace(',', '')
autos['price'] = autos['price'].astype(float)

autos['price'].head()

0    5000.0
1    8500.0
2    8990.0
3    4350.0
4    1350.0
Name: price, dtype: float64

In [19]:
# Clean and convert `odometer` column to float type
autos['odometer'] = autos['odometer'].str.replace('km', '')
autos['odometer'] = autos['odometer'].str.replace(',', '')
autos['odometer'] = autos['odometer'].astype(float)

# Rename `odometer` to `odometer_km`
autos.rename({'odometer':'odometer_km'}, axis=1, inplace=True)

autos['odometer_km'].head()

0    150000.0
1    150000.0
2     70000.0
3     70000.0
4    150000.0
Name: odometer_km, dtype: float64

In [20]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 17 columns):
date_crawled             50000 non-null object
name                     50000 non-null object
price                    50000 non-null float64
abtest                   50000 non-null object
vehicle_type             44905 non-null object
year_of_registration     50000 non-null int64
gearbox                  47320 non-null object
power_ps                 50000 non-null int64
model                    47242 non-null object
odometer_km              50000 non-null float64
month_of_registration    50000 non-null int64
fuel_type                45518 non-null object
brand                    50000 non-null object
not_repaired_damage      40171 non-null object
date_created             50000 non-null object
postal_code              50000 non-null int64
last_seen                50000 non-null object
dtypes: float64(2), int64(4), object(11)
memory usage: 6.5+ MB


### Exploring `price` and `odometer_km`
Since these features can take on a large range of values, I will explore them in a little more detail to determine if there are many outliers or anomalous data points.

In [21]:
print('Lowest price listed: $', autos['price'].min())
print('Most expensive price listed: $', autos['price'].max())

Lowest price listed: $ 0.0
Most expensive price listed: $ 99999999.0


In [22]:
print('Number of unique prices listed:', autos['price'].unique().shape[0], '\n')
print(autos['price'].describe())

Number of unique prices listed: 2357 

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


In [23]:
autos['price'].value_counts().sort_index(ascending=True).head(50)

0.0      1421
1.0       156
2.0         3
3.0         1
5.0         2
8.0         1
9.0         1
10.0        7
11.0        2
12.0        3
13.0        2
14.0        1
15.0        2
17.0        3
18.0        1
20.0        4
25.0        5
29.0        1
30.0        7
35.0        1
40.0        6
45.0        4
47.0        1
49.0        4
50.0       49
55.0        2
59.0        1
60.0        9
65.0        5
66.0        1
70.0       10
75.0        5
79.0        1
80.0       15
89.0        1
90.0        5
99.0       19
100.0     134
110.0       3
111.0       2
115.0       2
117.0       1
120.0      39
122.0       1
125.0       8
129.0       1
130.0      15
135.0       1
139.0       1
140.0       9
Name: price, dtype: int64

In [24]:
autos[autos['price'] == 100].head()

date_crawled                                           name  \
461   2016-03-17 12:54:04                        Gut_erhaltene_Alufelgen   
1097  2016-03-15 20:48:32                                   Ford_ka_!!!!   
1432  2016-03-07 13:52:40                         Ford_Escort_90PS_Kombi   
2557  2016-03-22 20:54:12                      Peugeot_106__Totalschaden   
2619  2016-03-20 00:57:37  Ford_fiesta_zum_ausschlachten_oder_herrichten   

      price   abtest vehicle_type  year_of_registration  gearbox  power_ps  \
461   100.0     test          NaN                  2008      NaN         0   
1097  100.0     test          NaN                  2000  manuell        60   
1432  100.0  control        kombi                  1995  manuell        90   
2557  100.0  control   kleinwagen                  2000  manuell         0   
2619  100.0  control   kleinwagen                  2000  manuell         0   

        model  odometer_km  month_of_registration fuel_type           brand  \
461       NaN      30000.0                      0       NaN  sonstige_autos   
1097       ka     150000.0                      0    benzin            ford   
1432   escort     150000.0                      1    benzin            ford   
2557  1_reihe     150000.0                      0    benzin         peugeot   
2619   fiesta     150000.0                     12    benzin            ford   

     not_repaired_damage         date_created  postal_code  \
461                  NaN  2016-03-17 00:00:00        66440   
1097                 NaN  2016-03-15 00:00:00        54290   
1432                  ja  2016-03-07 00:00:00        95497   
2557                  ja  2016-03-22 00:00:00        15234   
2619                 NaN  2016-03-19 00:00:00        38114   

                last_seen  
461   2016-03-27 00:16:38  
1097  2016-04-07 03:45:23  
1432  2016-03-17 17:15:34  
2557  2016-03-22 20:54:12  
2619  2016-04-06 02:17:23

In [25]:
autos['price'].value_counts().sort_index(ascending=False).head(20)

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
220000.0      1
198000.0      1
197000.0      1
Name: price, dtype: int64

In [26]:
autos[autos['price'] == 999990]

date_crawled                 name     price abtest vehicle_type  \
37585  2016-03-29 11:38:54  Volkswagen_Jetta_GT  999990.0   test    limousine   

       year_of_registration  gearbox  power_ps  model  odometer_km  \
37585                  1985  manuell       111  jetta     150000.0   

       month_of_registration fuel_type       brand not_repaired_damage  \
37585                     12    benzin  volkswagen                  ja   

              date_created  postal_code            last_seen  
37585  2016-03-29 00:00:00        50997  2016-03-29 11:38:54

In [27]:
autos[autos['price'] == 350000]

date_crawled         name     price   abtest vehicle_type  \
36818  2016-03-27 18:37:37  Porsche_991  350000.0  control        coupe   

       year_of_registration  gearbox  power_ps model  odometer_km  \
36818                  2016  manuell       500   911       5000.0   

       month_of_registration fuel_type    brand not_repaired_damage  \
36818                      3    benzin  porsche                nein   

              date_created  postal_code            last_seen  
36818  2016-03-27 00:00:00        70499  2016-03-27 18:37:37

### Where to Cut the Tails: `price`

At the low end of the spectrum there appear to be dozen of car for sale under \$100 which seems extremely cheap for even an old junky car. There seems to be a cluster at \$100 so that seems like a "reasonable" low-end price point and examining some of those listed for \$100, it appears a lot of those cars are rather old and have many kilometers on the engine. 

At the high end, there are a few cars listed for over a million dollars (and some much higher). Those prices seem unrealistic. Luckily there appears to be a nice break in the data between \$999,990 and the next highest price at \$350,000. Examine both of these listings individually it appears that the listing for \$350,000 might be a steep price for a Porsche 911, it does appear to be a new-ish model with very few kilometers on the engine. On the contrast, the listing for \$999990 for a Volkswagen Jetta from 1985 with 150,000km on the engine seems unrealistic. 

I conclude that I will snip the tails at:
\$100 <= `price` >= \$350,000

In [28]:
print('Lowest odometer reading listed: {} km'.format(autos['odometer_km'].min()))
print('Highest odometer reading listed: {} km'.format(autos['odometer_km'].max()))

Lowest odometer reading listed: 5000.0 km
Highest odometer reading listed: 150000.0 km


In [29]:
print('Number of unique odometer readings listed:', autos['odometer_km'].unique().shape[0], '\n')
print(autos['odometer_km'].describe())

Number of unique odometer readings listed: 13 

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


In [30]:
autos['odometer_km'].value_counts(dropna=False).sort_index()

5000.0        967
10000.0       264
20000.0       784
30000.0       789
40000.0       819
50000.0      1027
60000.0      1164
70000.0      1230
80000.0      1436
90000.0      1757
100000.0     2169
125000.0     5170
150000.0    32424
Name: odometer_km, dtype: int64

### Where to Cut the Tails: `odometer_km`
There are only 13 unique odometer readings for all of the listed cars and a vast majority of which are 150,000 km. It appears that users may have had to select an odometer reading bucket when creating a posting since all of the values fall nice round values. There may not have been an option for odometer readings above 150,000 km which is why so many users selected the highest bucket or perhaps on the website, that bucket was listed as 150,000 km or greater.

Whatever the case, there doesn't appear to be any outliers (or even any missing values).

In [31]:
# keeping only rows where the price lies between $100 and $350,000 (inclusive)
autos = autos[autos['price'].between(100, 350000)]

In [32]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48224 entries, 0 to 49999
Data columns (total 17 columns):
date_crawled             48224 non-null object
name                     48224 non-null object
price                    48224 non-null float64
abtest                   48224 non-null object
vehicle_type             43801 non-null object
year_of_registration     48224 non-null int64
gearbox                  46019 non-null object
power_ps                 48224 non-null int64
model                    45829 non-null object
odometer_km              48224 non-null float64
month_of_registration    48224 non-null int64
fuel_type                44345 non-null object
brand                    48224 non-null object
not_repaired_damage      39340 non-null object
date_created             48224 non-null object
postal_code              48224 non-null int64
last_seen                48224 non-null object
dtypes: float64(2), int64(4), object(11)
memory usage: 6.6+ MB


### Exploring `date_crawled`, `date_created` and `last_seen`
These columns are encoded as strings but represent a date (and time). The first 10 characters in the string represent the date component and remaining characters represent the time component (if applicable). 

The date format is: YYYY-MM-DD

In [33]:
autos['date_crawled'].str[0:10].unique().shape[0]

34

In [34]:
autos['date_crawled'].str[0:10].value_counts(dropna=False).sort_index()

2016-03-05    1223
2016-03-06     677
2016-03-07    1739
2016-03-08    1600
2016-03-09    1592
2016-03-10    1557
2016-03-11    1572
2016-03-12    1780
2016-03-13     756
2016-03-14    1768
2016-03-15    1655
2016-03-16    1421
2016-03-17    1519
2016-03-18     622
2016-03-19    1675
2016-03-20    1823
2016-03-21    1794
2016-03-22    1586
2016-03-23    1557
2016-03-24    1420
2016-03-25    1519
2016-03-26    1558
2016-03-27    1501
2016-03-28    1686
2016-03-29    1645
2016-03-30    1627
2016-03-31    1536
2016-04-01    1625
2016-04-02    1717
2016-04-03    1862
2016-04-04    1762
2016-04-05     630
2016-04-06     153
2016-04-07      67
Name: date_crawled, dtype: int64

### Distribution of `date_crawled`
The dataset was crawled from the eBay website all in the span of about a month, from the first week of March 2016 until the first week of April 2016, a range of 34 days. Except for a few dates, the distribution appears fairly uniform which means that each date, there were about the same number of results add to the dataset per day (around ~1600/day).

In [35]:
autos['date_created'].str[0:10].unique().shape[0]

76

In [36]:
autos['date_created'].str[0:10].describe()

count          48224
unique            76
top       2016-04-03
freq            1874
Name: date_created, dtype: object

In [37]:
print('Earliest date post was created:', autos['date_created'].str[0:10].min())
print('Latest date post was created:', autos['date_created'].str[0:10].max())

Earliest date post was created: 2015-06-11
Latest date post was created: 2016-04-07


In [38]:
autos[autos['date_created'].str[0:10] == '2016-04-07'].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 60 entries, 361 to 49569
Data columns (total 17 columns):
date_crawled             60 non-null object
name                     60 non-null object
price                    60 non-null float64
abtest                   60 non-null object
vehicle_type             55 non-null object
year_of_registration     60 non-null int64
gearbox                  55 non-null object
power_ps                 60 non-null int64
model                    57 non-null object
odometer_km              60 non-null float64
month_of_registration    60 non-null int64
fuel_type                53 non-null object
brand                    60 non-null object
not_repaired_damage      47 non-null object
date_created             60 non-null object
postal_code              60 non-null int64
last_seen                60 non-null object
dtypes: float64(2), int64(4), object(11)
memory usage: 8.4+ KB


### What About `date_created`?
Date created refers to the date that the post was made on the eBay website. The earliest post was created on May 11th, 2015 and the last post was created on April 7th, 2016 which also happens to be the last date the data was crawled. This makes sense. The data crawled appears to be collected from post made about 11 months prior the last day the data crawled extracted posts. 

In [39]:
autos['last_seen'].str[0:10].unique().shape[0]

34

In [40]:
autos['last_seen'].str[0:10].value_counts(dropna=False).sort_index()

2016-03-05       52
2016-03-06      208
2016-03-07      262
2016-03-08      353
2016-03-09      462
2016-03-10      513
2016-03-11      598
2016-03-12     1147
2016-03-13      428
2016-03-14      609
2016-03-15      765
2016-03-16      793
2016-03-17     1355
2016-03-18      353
2016-03-19      760
2016-03-20      996
2016-03-21      991
2016-03-22     1030
2016-03-23      896
2016-03-24      953
2016-03-25      921
2016-03-26      804
2016-03-27      750
2016-03-28     1005
2016-03-29     1075
2016-03-30     1191
2016-03-31     1149
2016-04-01     1102
2016-04-02     1200
2016-04-03     1212
2016-04-04     1183
2016-04-05     6031
2016-04-06    10704
2016-04-07     6373
Name: last_seen, dtype: int64

### And What About `last_seen`?

The `last_seen` feature refers to the last time the data crawler saw a specific post and it make sense that the date range would align with the date range from `date_crawled`. Now to ask, is this information important? 

### Now on to `year_of_registration`...

The `year_of_registration` feature captures when the car was first registered which means the year should take on values no later than 2016, when this dataset was first compiled.  Cars with years of registration earlier than the 1920s (or 1900 at the very earliest) should also be a viewed as possible anomalies. 

In [41]:
autos['year_of_registration'].describe()

count    48224.000000
mean      2004.730964
std         87.897388
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: year_of_registration, dtype: float64

In [42]:
autos[autos['year_of_registration'] >= 2017]['year_of_registration'].describe()

count    1867.00000
mean     2050.97429
std       441.16511
min      2017.00000
25%      2017.00000
50%      2017.00000
75%      2018.00000
max      9999.00000
Name: year_of_registration, dtype: float64

In [43]:
autos[autos['year_of_registration'] <= 1920]['year_of_registration'].describe()

count       7.000000
mean     1504.571429
std       440.903564
min      1000.000000
25%      1056.000000
50%      1800.000000
75%      1855.000000
max      1910.000000
Name: year_of_registration, dtype: float64

In [44]:
autos[autos['year_of_registration'].between(1000,1920)]

date_crawled                                   name   price  \
10556  2016-04-01 06:02:10                             UNFAL_Auto   450.0   
22316  2016-03-29 16:56:41  VW_Kaefer.__Zwei_zum_Preis_von_einem.  1500.0   
22659  2016-03-14 08:51:18                           Opel_Corsa_B   500.0   
24511  2016-03-17 19:45:11            Trabant__wartburg__Ostalgie   490.0   
28693  2016-03-22 17:48:41                         Renault_Twingo   599.0   
32585  2016-04-02 16:56:39                             UNFAL_Auto   450.0   
49283  2016-03-15 18:38:53                             Citroen_HY  7750.0   

        abtest vehicle_type  year_of_registration  gearbox  power_ps   model  \
10556  control          NaN                  1800      NaN      1800     NaN   
22316  control          NaN                  1000  manuell         0  kaefer   
22659     test          NaN                  1910      NaN         0   corsa   
24511  control          NaN                  1111      NaN         0     NaN   
28693  control   kleinwagen                  1910  manuell         0     NaN   
32585  control          NaN                  1800      NaN      1800     NaN   
49283  control          NaN                  1001      NaN         0  andere   

       odometer_km  month_of_registration fuel_type       brand  \
10556       5000.0                      2       NaN  mitsubishi   
22316       5000.0                      0    benzin  volkswagen   
22659     150000.0                      0       NaN        opel   
24511       5000.0                      0       NaN     trabant   
28693       5000.0                      0    benzin     renault   
32585       5000.0                      2       NaN  mitsubishi   
49283       5000.0                      0       NaN     citroen   

      not_repaired_damage         date_created  postal_code  \
10556                nein  2016-04-01 00:00:00        63322   
22316                 NaN  2016-03-29 00:00:00        48324   
22659                 NaN  2016-03-14 00:00:00        52393   
24511                 NaN  2016-03-17 00:00:00        16818   
28693                 NaN  2016-03-22 00:00:00        70376   
32585                nein  2016-04-02 00:00:00        63322   
49283                 NaN  2016-03-15 00:00:00        66706   

                 last_seen  
10556  2016-04-01 09:42:30  
22316  2016-03-31 10:15:28  
22659  2016-04-03 07:53:55  
24511  2016-04-07 07:17:29  
28693  2016-04-06 09:16:59  
32585  2016-04-04 14:46:21  
49283  2016-04-06 18:47:20

### Cleaning Up `year_of_registration`

From the analysis above, there are a handful of cars whose year of registration fall are after the car was posted, 2016. Any car whose registration year is 2017 or later will be removed.

As for cars whose year of registration is before 1920, of which there are only 7, also appear save to remove. These 4 cars listed with a year of registration in 1910 are for cars which did yet exist (Opel and Renault). The remaining 5 cars from this list have registration years far before cars were even invented an can also be safely removed from the dataset.

In [45]:
# keeping only rows where the year_of_registration lies 
# between 1920 and 2016 (inclusive)
autos = autos[autos['year_of_registration'].between(1920, 2016)]

In [46]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 46350 entries, 0 to 49999
Data columns (total 17 columns):
date_crawled             46350 non-null object
name                     46350 non-null object
price                    46350 non-null float64
abtest                   46350 non-null object
vehicle_type             43798 non-null object
year_of_registration     46350 non-null int64
gearbox                  44377 non-null object
power_ps                 46350 non-null int64
model                    44220 non-null object
odometer_km              46350 non-null float64
month_of_registration    46350 non-null int64
fuel_type                43180 non-null object
brand                    46350 non-null object
not_repaired_damage      38254 non-null object
date_created             46350 non-null object
postal_code              46350 non-null int64
last_seen                46350 non-null object
dtypes: float64(2), int64(4), object(11)
memory usage: 6.4+ MB


In [55]:
autos['brand'].value_counts()

volkswagen        9799
bmw               5107
opel              4970
mercedes_benz     4480
audi              4022
ford              3237
renault           2181
peugeot           1384
fiat              1187
seat               846
skoda              761
nissan             711
mazda              706
smart              658
citroen            651
toyota             593
hyundai            464
sonstige_autos     442
volvo              423
mini               408
mitsubishi         379
honda              365
kia                328
alfa_romeo         309
porsche            279
suzuki             275
chevrolet          263
chrysler           163
dacia              123
daihatsu           116
jeep               106
land_rover          98
subaru              98
saab                77
jaguar              71
daewoo              69
trabant             63
rover               62
lancia              49
lada                27
Name: brand, dtype: int64

## Top Posted Car Brands

I would like to explore the top 25% of car brands posted. These top brands seem to account for a vast majority of the cars posted on the website. 

In [127]:
# number of unique brands in the dataset
nbrands = autos['brand'].nunique()
# calculate what number corresponds to 25% of the number of car brands
top_brands = int(nbrands/4)
# extract the  top 25% of the brands by name into a list
brands = autos['brand'].value_counts().head(top_brands).index

# calculate the average price for the top brands and store them in a dict 
# where the key is the brand, and the value is the average price across that brand
top_brands_prices = {}
for b in brands:
    mean_price_brand = autos[autos['brand'] == b]['price'].mean()
    if b not in top_brands_prices:
        top_brands_prices[b] = round(mean_price_brand,2)

In [128]:
top_brands_prices

{'audi': 9380.72,
 'bmw': 8381.68,
 'fiat': 2836.87,
 'ford': 3779.27,
 'mercedes_benz': 8672.65,
 'opel': 3006.0,
 'peugeot': 3113.86,
 'renault': 2496.94,
 'seat': 4433.42,
 'volkswagen': 5436.95}

### Average Price: Top Posted Car Brands

There appears to be some segregation amoung the top posted car brands. At the high end, car brands such as Audi, BMW, and Mercedes-Benz seem to be listed at prices above \$8300. On the low end, car brands such as Fiat, Ford, Opel, Peugeot and Renault seem to be listed at around \$3800 and below. And lastly, car brands Seat and Volkswagen seem to hover somewhere in the middle price range.

In [131]:
top_brands_mileage = {}
for b in brands:
    mean_mileage_brand = autos[autos['brand'] == b]['odometer_km'].mean()
    if b not in top_brands_mileage:
        top_brands_mileage[b] = round(mean_mileage_brand)

In [132]:
top_brands_mileage

{'audi': 129245.0,
 'bmw': 132695.0,
 'fiat': 116950.0,
 'ford': 124277.0,
 'mercedes_benz': 131026.0,
 'opel': 129380.0,
 'peugeot': 127128.0,
 'renault': 128338.0,
 'seat': 121537.0,
 'volkswagen': 128800.0}

In [133]:
brands_prices = pd.Series(top_brands_prices)
brands_mileage = pd.Series(top_brands_mileage)

brands = pd.DataFrame(brands_prices, columns=['mean_price'])
brands['mean_mileage'] = brands_mileage

In [134]:
brands

mean_price  mean_mileage
audi              9380.72      129245.0
bmw               8381.68      132695.0
fiat              2836.87      116950.0
ford              3779.27      124277.0
mercedes_benz     8672.65      131026.0
opel              3006.00      129380.0
peugeot           3113.86      127128.0
renault           2496.94      128338.0
seat              4433.42      121537.0
volkswagen        5436.95      128800.0

### Average Mileage: Top Posted Car Brands

It appears that all of the top 10 car brands have roughly the same average mileage, all well over 110,000km. Fiat has the lowest average with roughly 117,000km whereas BMW has the highest average with roughly 133,000km.

